### PR TITLE
add fleet test tools.

### PR DESCRIPTION
**Pull request by wuhuachaocoding**
### PR types

New features
### PR changes

 APIs
### Description

Card-73481
add fleet test tools: update and add perf test APIs.
- fleet.perf_test: test various communications perf using various communication package.
- fleet.monitor_perf: monitor specific communication perf using various communication package.

### CODE REVIEW

File: python/paddle/distributed/fleet/fleet.py [hunk 1]
```diff
@@ -382,9 +382,11 @@ def init(
                 )
         return self
 
-    def perf_test(self, round=50):
+    def perf_test(self, round=50, test_comm=[], context={}, hcg=None):
         # test allreduce perf
         def allreduce_test(iteration, x, group):
+            if group is None or group.nranks <= 1:
+                return -1
```
Comment (gongweibao) [inline python/paddle/distributed/fleet/fleet.py]: log WARNING?

File: python/paddle/distributed/fleet/fleet.py [hunk 2]
```diff
@@ -420,42 +426,168 @@ def broadcast_test(iteration, x, group):
...
             )
-            ret = broadcast_test(iteration=round, x=x, group=test_group)
-            logger.info(
-                f"[BroadcastTest] nbytes {nbytes}B test result: {ret} s/iter"
+            reduce_scatter_size, reduce_scatter_thres_time = context.get(
+                "reduce_scatter", [nbytes, -1]
             )
+
+            # inter machines
+            if "allreduce" in test_comm:
```
Comment (gongweibao) [inline python/paddle/distributed/fleet/fleet.py:578]: Split this large function into small ones?

**Review by gongweibao — CHANGES_REQUESTED** [review]

File: python/paddle/distributed/fleet/fleet.py [hunk 3]
```diff
@@ -382,9 +382,11 @@ def init(
                 )
         return self
 
-    def perf_test(self, round=50):
+    def perf_test(self, round=50, test_comm=[], context={}, hcg=None):
         # test allreduce perf
         def allreduce_test(iteration, x, group):
```
Comment (gongweibao) [inline python/paddle/distributed/fleet/fleet.py]: 不用在函数中嵌入式的定义函数？

File: python/paddle/fluid/tests/unittests/hybrid_parallel_perf_test.py [hunk 4]
```diff
@@ -60,6 +60,106 @@ def build_optimizer(self, model):
...
+                "reduce",
+                "broadcast",
+                "allgather",
+                "reduce_scatter",
+            ],
+        )
+        # context: {comm_type:[size, time]}
+        # only test allreduce for package(1024B) and time threshold(0.00000001s),
+        # and test allgather for package(8192B) and time threshold(2s),
+        fleet.perf_test(
```
Comment (gongweibao) [inline python/paddle/fluid/tests/unittests/hybrid_parallel_perf_test.py:77]: 返回值是什么？

File: python/paddle/distributed/fleet/fleet.py (hunk 2 above)
Comment (wuhuachaocoding) [reply python/paddle/distributed/fleet/fleet.py:578]: DONE

File: python/paddle/fluid/tests/unittests/hybrid_parallel_perf_test.py (hunk 4 above)
Comment (wuhuachaocoding) [reply python/paddle/fluid/tests/unittests/hybrid_parallel_perf_test.py:77]: 这里不需要返回值，测试结果都打印出来了

File: python/paddle/distributed/fleet/fleet.py (hunk 1 above)
Comment (wuhuachaocoding) [reply python/paddle/distributed/fleet/fleet.py]: DONE

File: python/paddle/distributed/fleet/fleet.py (hunk 3 above)
Comment (wuhuachaocoding) [reply python/paddle/distributed/fleet/fleet.py]: DONE

**Review by ForFishes — APPROVED** [review]
LGTM

**Review by gongweibao — APPROVED** [review]
LGTM

**Pull request merged**